### ISSUE TITLE
Void linux support for mach

### ISSUE THREAD

**Issue by mamins1376** [servo/servo#26114]
I'm trying to build on void linux. Already added the packages to `./mach bootstrap`, but no luck with building. Does this have anything to do with the distro or is it another issue for cargo itself?
```sh
$ ./mach build
error: dep-info invalid
note: this is an unexpected cargo internal error
note: we would appreciate a bug report: https://github.com/rust-lang/cargo/issues/
note: cargo 1.43.0-nightly (bda50510d 2020-03-02)
Build FAILED in 0:00:07
```
I also tried with `RUST_LOG=cargo=debug`, but no difference in output. How can I get more information?

**Comment by SimonSapin** [comment]
Cargo uses `CARGO_LOG` instead of `RUST_LOG`.

**Comment by mamins1376** [comment]
Turns out cargo panic was due to outdated objects in my `target/`. Removing those solved the issue.
Now I'm stuck with building `components/net`. it's `build.rs` checks for `DEP_OPENSSL_VERSION_NUMBER` (https://github.com/servo/servo/commit/21b2479dc27221c6cf0340210de6a6bd656ab01f) but *somehow* it is not set.
Void linux uses libressl instead of openssl. Who should set that variable?

**Comment by SimonSapin** [comment]
`DEP_OPENSSL_VERSION_NUMBER` is set by Cargo https://doc.rust-lang.org/cargo/reference/build-scripts.html#the-links-manifest-key from the value printed here:

* https://github.com/sfackler/rust-openssl/blob/b027f1603189919d5f63c6aff483243aaa188568/openssl-sys/build/main.rs#L228
* https://github.com/sfackler/rust-openssl/blob/b027f1603189919d5f63c6aff483243aaa188568/openssl-sys/Cargo.toml#L11

It looks like `components/net/build.rs` should check for `DEP_OPENSSL_LIBRESSL_VERSION_NUMBER` as well, since that’s what the print in the other branch would set:

* https://github.com/sfackler/rust-openssl/blob/b027f1603189919d5f63c6aff483243aaa188568/openssl-sys/build/main.rs#L196

Do you know what LibreSSL versions are expected to be API-compatible with OpenSSL 1.1.1? CC @Darkspirit who added this check in https://github.com/servo/servo/pull/25736

**Comment by Darkspirit** [comment]
LibreSSL ([description](https://wiki.gentoo.org/wiki/Project:LibreSSL)) is almost a drop-in replacement for outdated versions of OpenSSL. 
LibreSSL does not support necessary TLS 1.3 and DTLS 1.2. Servo shouldn't have to maintain support for three entirely different TLS libraries at the end.

Gstreamer's DTLS implementation does not support LibreSSL:
void linux applies [their own patch](https://github.com/void-linux/void-packages/blob/c44ab9a2e2ea75fd7effbf38f43a75ca35832f55/srcpkgs/gst-plugins-bad1/patches/libressl.patch) to enforce the [code path](https://gitlab.freedesktop.org/gstreamer/gst-plugins-bad/-/blob/f10b424418e448211e3427a76fcd046e157ef0b7/ext/dtls/gstdtlsagent.c#L185) of [deprecated](https://bugs.chromium.org/p/webrtc/issues/detail?id=10261#c74) DTLS 1.0, meant for outdated versions of OpenSSL. 
Firefox Nightly [requires](https://bugzilla.mozilla.org/show_bug.cgi?id=1506392) at minimum DTLS 1.2. DTLS 1.0 removal almost shipped from Beta to Stable, but was slightly postponed because of Covid19. (The upcoming release of Jitsi Meet will finally have DTLS 1.2 support and unblock removal.)

With #24764 I suggested to introduce a Rustls build option, as statically linked TLS library. It was postponed at that time because of understandable circumstances.
The Rust OpenSSL package also offers a "vendored" feature flag which makes use of https://github.com/alexcrichton/openssl-src-rs. This could be used for Void Linux (or for Linux in general #20783). Either with using Void Linux' _sabotaged_ Gstreamer or a statically linked one.

**Comment by mamins1376** [comment]
Is automatically enabling `vendored` flag by mach on void linux an acceptable behavior?
